### PR TITLE
*: record more rpc runtime information in cop runtime stats

### PR DESCRIPTION
**Pull request by crazycs520**
<!-- Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?

Record more RPC runtime information in cop runtime stats, such as:
* send rpc request count and type
* backoff information

Here is some example:

**Normally**

```sql
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+--------------+------+
| id                     | estRows  | actRows | task      | access object | execution info                                                                                                                                                                                                                     | operator info                   | memory       | disk |
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+--------------+------+
| StreamAgg_20           | 1.00     | 1       | root      |               | time:177.902235ms, loops:2                                                                                                                                                                                                         | funcs:count(Column#9)->Column#4 | 9.484375 KB  | N/A  |
| └─TableReader_21       | 1.00     | 1024    | root      |               | time:177.870315ms, loops:2, cop_task: {num: 1024, max: 21.191093ms, min: 217.89µs, avg: 1.831162ms, p95: 6.302146ms, max_proc_keys: 83, p95_proc_keys: 72, tot_proc: 100ms, tot_wait: 41ms, rpc_num: 1024, rpc_time: 1.870528822s} | data:StreamAgg_8                | 2.3359375 KB | N/A  |
|   └─StreamAgg_8        | 1.00     | 1024    | cop[tikv] |               | proc max:1ms, min:0s, p80:0s, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                      | funcs:count(1)->Column#9        | N/A          | N/A  |
|     └─TableFullScan_18 | 10000.00 | 60236   | cop[tikv] | table:t1      | proc max:1ms, min:0s, p80:0s, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                      | keep order:false, stats:pseudo  | N/A          | N/A  |
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+--------------+------+
```

**When 1 tikv-server is offline**

```sql
+------------------------+-----------+---------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| id                     | estRows   | actRows | task      | access object | execution info                                                                                                                                                                                                                                                                            | operator info                   | memory         | disk |
+------------------------+-----------+---------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| StreamAgg_20           | 1.00      | 1       | root      |               | time:9.004457817s, loops:2                                                                                                                                                                                                                                                                | funcs:count(Column#9)->Column#4 | 9.484375 KB    | N/A  |
| └─TableReader_21       | 1.00      | 1024    | root      |               | time:9.004429257s, loops:2, cop_task: {num: 1024, max: 2.99937716s, min: 436.349µs, avg: 27.14526ms, p95: 1.246608ms, max_proc_keys: 613, p95_proc_keys: 551, tot_proc: 321ms, tot_wait: 146ms, rpc_num: 1043, rpc_time: 27.791302111s}, backoff{regionMiss: 11.168s, tikvRPC: 1m49.042s} | data:StreamAgg_8                | 1.435546875 KB | N/A  |
|   └─StreamAgg_8        | 1.00      | 1024    | cop[tikv] |               | proc max:1ms, min:0s, p80:1ms, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                                                                            | funcs:count(1)->Column#9        | N/A            | N/A  |
|     └─TableFullScan_18 | 453621.00 | 524910  | cop[tikv] | table:t1      | proc max:1ms, min:0s, p80:1ms, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                                                                            | keep order:false, stats:pseudo  | N/A            | N/A  |
+------------------------+-----------+---------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
```

### What is changed and how it works?


### Related changes

- Need to cherry-pick to the release branch

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note <!-- bugfixes or new feature need a release note -->

- Record more RPC runtime information in cop runtime stats

### CODE REVIEW

File: distsql/select_result.go [hunk 1]
```diff
@@ -297,3 +318,95 @@ func (r *selectResult) Close() error {
...
+	if s.totalProcessTime > 0 {
+		buf.WriteString(", ")
+		buf.WriteString(fmt.Sprintf("Coprocessor: {total_process_time: %s", s.totalProcessTime.String()))
+		if s.totalWaitTime > 0 {
+			buf.WriteString(", ")
+			buf.WriteString("total_wait_time: ")
+			buf.WriteString(s.totalWaitTime.String())
+		}
+		buf.WriteString("}")
+	}
```
Comment (qw4990) [inline distsql/select_result.go:399]: Can we put these fields into `cop_task` like `cop_task : {tot_proc_t: ..., tot_wait_t: ..., num: ..., max: ...}`?

File: distsql/select_result.go [hunk 2]
```diff
@@ -297,3 +318,95 @@ func (r *selectResult) Close() error {
...
+				sum += float64(t)
+			}
+			vAvg := time.Duration(sum / float64(size))
+
+			sort.Slice(s.procKeys, func(i, j int) bool {
+				return s.procKeys[i] < s.procKeys[j]
+			})
+			keyMax := s.procKeys[size-1]
+			keyP95 := s.procKeys[size*19/20]
+			buf.WriteString(fmt.Sprintf("cop_task: {num: %v, max :%v, min:%v, avg:%v, p80:%v, p95:%v}, proc_keys: {max:%v, p95:%v}", size, vMax, vMin, vAvg, vP80, vP95, keyMax, keyP95))
```
Comment (qw4990) [inline distsql/select_result.go:380]: How about merging these two fields into one? 
For example:
```
cop_task: {num: %v, max :%v, min:%v, avg:%v, p80:%v, p95:%v, max_keys: %v, p95_keys: %v}
```

File: distsql/select_result.go [hunk 3]
```diff
@@ -297,3 +318,95 @@ func (r *selectResult) Close() error {
...
+	}
+
+	if len(s.backoffSleep) > 0 {
+		buf.WriteString(", ")
+		for k, v := range s.backoffTimes {
+			if buf.Len() > 0 {
+				buf.WriteByte(',')
+			}
+			d := s.backoffSleep[k]
+			buf.WriteString(fmt.Sprintf("%s_backoff:{num:%d, total_time:%s}", k, v, d.String()))
```
Comment (qw4990) [inline distsql/select_result.go:408]: How about combining them into one field: `backoff: {type1_num: ..., type1_time: ..., type2_num: ..., type2_time: ...}`?

**Comment by qw4990** [comment]
And please fix CI at your convenience @crazycs520 .

File: distsql/select_result.go (hunk 1 above)
Comment (crazycs520) [reply distsql/select_result.go]: great, done.

File: distsql/select_result.go (hunk 2 above)
Comment (crazycs520) [reply distsql/select_result.go]: great, done.

File: distsql/select_result.go (hunk 3 above)
Comment (crazycs520) [reply distsql/select_result.go:420]: Em... I still prefer the current display. @wjhuang2016 PTAL

File: distsql/select_result.go [hunk 4]
```diff
@@ -297,3 +320,106 @@ func (r *selectResult) Close() error {
 	}
 	return r.resp.Close()
 }
+
+// HasCopRuntimeStats is a interface uses to check whether the result has cop runtime stats.
+type HasCopRuntimeStats interface {
```
Comment (wjhuang2016) [inline distsql/select_result.go:325]: I don't think HasCopRuntimeStats is a good name. CopRuntimeStats maybe better.

**Comment by zz-jason** [comment]
> | └─TableReader_17       | 1.00      | 256     | root      |               | time:4.712924994s, loops:2, cop_task: {total_process_time: 84ms, total_wait_time: 81ms, total_process_keys: 102572, num: 256, max: 1.560532569s, min: 578.4µs, avg: 12.176646ms, p80: 862.04µs, p95: 995.561µs, max_process_keys: 465, p95_process_keys: 431}, Cop:{num_rpc:258, total_time:3.116192098s}, regionMiss_backoff:{num:343, total_time:3.908s}, tikvRPC_backoff:{num:89, total_time:55.872s} | data:StreamAgg_8                                               | 702 Bytes   | N/A  |

it's too long to display.

**Comment by crazycs520** [comment]
> > | └─TableReader_17       | 1.00      | 256     | root      |               | time:4.712924994s, loops:2, cop_task: {total_process_time: 84ms, total_wait_time: 81ms, total_process_keys: 102572, num: 256, max: 1.560532569s, min: 578.4µs, avg: 12.176646ms, p80: 862.04µs, p95: 995.561µs, max_process_keys: 465, p95_process_keys: 431}, Cop:{num_rpc:258, total_time:3.116192098s}, regionMiss_backoff:{num:343, total_time:3.908s}, tikvRPC_backoff:{num:89, total_time:55.872s} | data:StreamAgg_8                                               | 702 Bytes   | N/A  |
> 
> it's too long to display.

Yes, the is the result when 1 tikv-server offline, normally, the result shouldn't have `backoff` information.

But, the information is really too long, here is some solution:

1. how about use abbreviation, such as:

```sql
total_process_time -> proc_time
total_wait_time -> wait_time
total_process_keys -> proc_keys
max_process_keys -> max_proc_keys
``` 

2. Move some information to the first cop task execution info, such as:

```sql
total_process_time 
total_process_time
total_process_keys
max_process_keys
p95_process_keys
```

3. Remove some runtime information

@zz-jason @qw4990  PTAL

**Comment by qw4990** [comment]
I prefer to keep this information since in most cases there is no `backoff` field and the length is acceptable.
What's your opinion? @zz-jason

**Comment by zz-jason** [comment]
> I prefer to keep this information since in most cases there is no `backoff` field and the length is acceptable.
> What's your opinion? @zz-jason

even without `backoff`, the length is too long to display in one row in the MySQL client. can we separate it into several rows no more than a fixed column size, for example 80 or 120?

**Comment by zz-jason** [comment]
using abbrv is a workaround, I prefer:

```
total_process_time -> tot_proc
total_wait_time -> tot_wait
total_process_keys -> tot_proc_keys
max_process_keys -> max_proc_keys
```

BTW, what's the different between `tot_proc_keys` and `actRows`?

**Comment by crazycs520** [comment]
> using abbrv is a workaround, I prefer:
> 
> ```
> total_process_time -> tot_proc
> total_wait_time -> tot_wait
> total_process_keys -> tot_proc_keys
> max_process_keys -> max_proc_keys
> ```
> 
> BTW, what's the different between `tot_proc_keys` and `actRows`?

Good catch, `tot_proc_keys` and `actRows` should be same, we can remove tot_proc_keys.

**Comment by crazycs520** [comment]
> > I prefer to keep this information since in most cases there is no `backoff` field and the length is acceptable.
> > What's your opinion? @zz-jason
> 
> even without `backoff`, the length is too long to display in one row in the MySQL client. can we separate it into several rows no more than a fixed column size, for example 80 or 120?

you mean like this?

```sql
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| id                     | estRows  | actRows | task      | access object | execution info                                                                                                                                                   | operator info                   | memory         | disk |
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| StreamAgg_20           | 1.00     | 1       | root      |               | time:187.635186ms, loops:2                                                                                                                                       | funcs:count(Column#9)->Column#4 | 9.484375 KB    | N/A  |
| └─TableReader_21       | 1.00     | 1024    | root      |               | time:187.600876ms, loops:2, cop_task: {total_process_time: 212ms, total_wait_time: 35ms, total_process_keys: 252352, num: 1024, max: 11.820529ms, min: 353.491µs | data:StreamAgg_8                | 2.392578125 KB | N/A  |
|                        |          |         |           |               |  avg: 1.804347ms, p80: 2.823734ms, p95: 5.7491ms, max_process_keys: 290, p95_process_keys: 273}, Cop:{num_rpc:1024, total_time:1.84294213s}                      |                                 |                |      |
|   └─StreamAgg_8        | 1.00     | 1024    | cop[tikv] |               | proc max:1ms, min:0s, p80:1ms, p95:1ms, iters:1024, tasks:1024                                                                                                   | funcs:count(1)->Column#9        | N/A            | N/A  |
|     └─TableFullScan_18 | 10000.00 | 252352  | cop[tikv] | table:t1      | proc max:1ms, min:0s, p80:1ms, p95:1ms, iters:1024, tasks:1024                                                                                                   | keep order:false, stats:pseudo  | N/A            | N/A  |
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
```

Great, it is much better to see. @qw4990 PTAL

**Comment by crazycs520** [comment]
> > I prefer to keep this information since in most cases there is no `backoff` field and the length is acceptable.
> > What's your opinion? @zz-jason
> 
> even without `backoff`, the length is too long to display in one row in the MySQL client. can we separate it into several rows no more than a fixed column size, for example 80 or 120?

BTW, after removing unnecessary information and use the abbreviation, the length with backoff information is a little better:

```sql
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| id                     | estRows  | actRows | task      | access object | execution info                                                                                                                                                                                                                                                                                                                                             | operator info                   | memory         | disk |
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| StreamAgg_20           | 1.00     | 1       | root      |               | time:2.901180903s, loops:2                                                                                                                                                                                                                                                                                                                                 | funcs:count(Column#9)->Column#4 | 9.484375 KB    | N/A  |
| └─TableReader_21       | 1.00     | 1024    | root      |               | time:2.901152623s, loops:2, cop_task: {num: 1024, max: 2.164663ms, min: 419.211µs, avg: 810.592µs, p80: 939.091µs, p95: 1.135112ms, max_proc_keys: 242, p95_proc_keys: 220, tot_proc: 138ms, tot_wait: 397ms}, Cop:{num_rpc:1024, total_time:824.343785ms}, tikvRPC_backoff:{num:75, total_time:33.919s}, regionMiss_backoff:{num:1099, total_time:3.908s} | data:StreamAgg_8                | 1.642578125 KB | N/A  |
|   └─StreamAgg_8        | 1.00     | 1024    | cop[tikv] |               | proc max:1ms, min:0s, p80:0s, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                                                                                                                                              | funcs:count(1)->Column#9        | N/A            | N/A  |
|     └─TableFullScan_18 | 10000.00 | 201722  | cop[tikv] | table:t1      | proc max:1ms, min:0s, p80:0s, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                                                                                                                                              | keep order:false, stats:pseudo  | N/A            | N/A  |
+------------------------+----------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
```

**Comment by crazycs520** [comment]
@zz-jason @qw4990 PTAL

**Comment by zz-jason** [comment]
```
cop_task: {num: 1024, max: 2.164663ms, min: 419.211µs, avg: 810.592µs, p80: 939.091µs, p95: 1.135112ms, max_proc_keys: 242, p95_proc_keys: 220, tot_proc: 138ms, tot_wait: 397ms}
Cop:{num_rpc:1024, total_time:824.343785ms}, tikvRPC_backoff:{num:75, total_time:33.919s}, regionMiss_backoff:{num:1099, total_time:3.908s}
```

I think we can improve more:

- we can merge `cop_task` and `Cop`
- for process times, we can remove p80, p95 is sufficient
- for regionMiss_backoff, I think we can remove `num:1099`

**Comment by crazycs520** [comment]
> I think we can improve more:
> 
> * we can merge `cop_task` and `Cop`
> * for process times, we can remove p80, p95 is sufficient
> * for regionMiss_backoff, I think we can remove `num:1099`

Great, thanks, Now the explain analyze result should be:

```sql
+------------------------+-----------+---------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| id                     | estRows   | actRows | task      | access object | execution info                                                                                                                                                                                                                                                                            | operator info                   | memory         | disk |
+------------------------+-----------+---------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
| StreamAgg_20           | 1.00      | 1       | root      |               | time:9.004457817s, loops:2                                                                                                                                                                                                                                                                | funcs:count(Column#9)->Column#4 | 9.484375 KB    | N/A  |
| └─TableReader_21       | 1.00      | 1024    | root      |               | time:9.004429257s, loops:2, cop_task: {num: 1024, max: 2.99937716s, min: 436.349µs, avg: 27.14526ms, p95: 1.246608ms, max_proc_keys: 613, p95_proc_keys: 551, tot_proc: 321ms, tot_wait: 146ms, rpc_num: 1043, rpc_time: 27.791302111s}, backoff{regionMiss: 11.168s, tikvRPC: 1m49.042s} | data:StreamAgg_8                | 1.435546875 KB | N/A  |
|   └─StreamAgg_8        | 1.00      | 1024    | cop[tikv] |               | proc max:1ms, min:0s, p80:1ms, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                                                                            | funcs:count(1)->Column#9        | N/A            | N/A  |
|     └─TableFullScan_18 | 453621.00 | 524910  | cop[tikv] | table:t1      | proc max:1ms, min:0s, p80:1ms, p95:1ms, iters:1024, tasks:1024                                                                                                                                                                                                                            | keep order:false, stats:pseudo  | N/A            | N/A  |
+------------------------+-----------+---------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+------+
```

**Comment by crazycs520** [comment]
@qw4990 @zz-jason PTAL

**Pull request merged**